langserver/fs.go

```diff
@@ -26,42 +26,67 @@ func IsFileSystemRequest(method string) bool {
 		method == "textDocument/didSave"
 }
 
-func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) error {
+// HandleFileSystemRequest handles textDocument/did* requests. The path the
+// request is for is returned. true is returned if a file was modified.
+func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) (string, bool, error) {
 	span := opentracing.SpanFromContext(ctx)
 	h.Mu.Lock()
 	overlay := h.overlay
 	h.Mu.Unlock()
 
+	var before []byte
+	var beforeErr error
```

Comment:
It seems less than ideal that this state needs to get set before calling checkChanged.

I would recommend either making these parameters to checkChanged or better yet, making checkChanged take a URI and an operation closure and having checkChanged call readFile before/after the operation.